lib/helpers/lib/each-openapi-module.js

```diff
@@ -0,0 +1,202 @@
+        node.type ??
+        '(unknown)'
+    
+    const format = types[node.type]?.format ?? (val => val)
+    return format(type) 
+}
+    
+    
+// recursive function to walk the Table of Contents
+// NOTE: we are building up HTML here, because Asciidoc is hard to generate for something that's 
+// richly marked up and whitespace dependent.
+function process_toc (node) {
+    
+    // the Left of the JSON expression (except at the Root) is `key: `
+    // IF the related node has a Description OR Title, then link to it.
+    
+    const left = (node, path) =>
+        path.length ?
+            (node.description || node.title) ?
+                `<a href="#${path_id_href(path)}">${path[path.length-1]}</a>: `
+                : `${path[path.length-1]}: `
+            : ''
+
+    const level = (path) => indent.repeat(path.length)
+    
+    const handleObject = (node, path, children) =>
+        level(path) + left(node, path) + '{\n' +
+        children.join(',\n') + '\n' +
+        level(path) + '}'
+
+    const handleArray = (node, path, children) =>
+        level(path) + left(node, path) + '[' +
+        (children.length > 1 ?
+            '\n' + children.join(',\n') + '\n' + level(path) :
+            children[0])
+        + ']'
+
+    const handleLeaf = (node, path) =>
+        level(path) + left(node, path) + format_type(node)
+
+    return iterate_properties(node, handleLeaf, handleObject, handleArray)
+}
+    
+// The usual path to get to child objects is `properties` which includes specific, *named* properties.
+//
+// `additionalProperties` are unrestrained collections.
+// They can also be specified as a boolean, which is a shorthand for a default object.
+
+function child_entries(node) {
+    const via_properties = 
+        node.properties ? 
+            Object.entries(node.properties) : []
+    
+    function format_additionalProperties (ap) {
+        if (typeof(ap) === 'boolean') ap = {}
+        return { type: "object", ...ap }
+    }
+    
+    const via_additionalProperties =
+        node.additionalProperties ?
+            [[
+                `{${node.additionalProperties["x-additionalPropertiesName"] || "additionalProperties"}...}`,
+                format_additionalProperties(node.additionalProperties) ]]
+            : []
+
+    const via_allOf = node.allOf?.flatMap(a => Object.entries(a.properties)) || []
+
+    return [...via_properties, ...via_additionalProperties, ...via_allOf]
+}
+
+// recursive function to walk the tree.
+// pass in callbacks for:
+//   `handleLeaf`
+//   `handleObject`
+//   `handleArray`
+// to handle the tree as appropriate
+// this can be used to:
+//    * generate a tree (as per process_toc)
+//    * flatten the tree into a list (as per get_items)
+function iterate_properties (node, handleLeaf, handleObject, handleArray) {
+    
+
+    function recurse(node, path) {
+
+        if (! node.type) {
+            if (
+                node?.additionalProperties?.type === 'object' || 
```

Comment:
I didn't know this before looking at it, but `additionalProperties` I think is always an object. It can actually be `additionalProperties: true` which is equivalent to `additionalProperties = {}` https://swagger.io/docs/specification/v3_0/data-models/dictionaries/#free-form-objects but that's not used in SG.

I think that this should be `node?.additionalProperties !== undefined`  rather than `node?.additionalProperties?.type === 'object'` to presume this is an object.